osf/utils/machines.py

```diff
@@ -143,3 +144,64 @@ def get_context(self):
             'provider_contact_email': self.machineable.provider.email_contact or 'contact@osf.io',
             'provider_support_email': self.machineable.provider.email_support or 'support@osf.io',
         }
+
+class RequestMachine(BaseMachine):
+    ActionClass = NodeRequestAction
+
+    def save_changes(self, ev):
+        """ Handles contributorship changes and state transitions
+        """
+        if ev.event.name == DefaultTriggers.ACCEPT.value:
+            self.machineable.target.add_contributor(self.machineable.creator, auth=Auth(ev.kwargs['user']), permissions=permissions.READ, send_email='{}_request'.format(self.machineable.request_type))
+        elif ev.event.name == DefaultTriggers.EDIT_COMMENT.value and self.action is not None:
+            self.machineable.comment = self.action.comment
+        self.machineable.save()
+
+    def resubmission_allowed(self, ev):
+        # TODO: [PRODUCT-395]
+        return False
+
+    def notify_submit(self, ev):
+        """ Notify admins that someone is requesting access
+        """
+        context = self.get_context()
+        context['contributors_url'] = '{}contributors/'.format(self.machineable.target.absolute_url)
+        context['project_settings_url'] = '{}settings/'.format(self.machineable.target.absolute_url)
+        for admin in self.machineable.target.admin_contributors:
+            mails.send_mail(
```

Comment:
just curious - seems like reviews uses signals to send emails, is there an advantage to sending the email directly here instead of sending a signal for consistency?

Reply:
Advantages include simplicity, readability, and transparency. There was a moderation bug where emails were not being sent -- this was because the signal listeners weren't imported before app initialization. The actual mail is sent asynchronously, so it shouldn't be any less efficient than using signals. 

The only case signals are really worth the hassle is when you need to trigger the same behavior from multiple locations where a method import would cause problems (e.g. circular imports). That isn't true here (and may or may not have been true for moderation), so I didn't use signals.

Reply:
Yup, makes sense as well!